### ISSUE TITLE
Ignore :global scope in selector-class-pattern

### ISSUE THREAD

**Issue by gpoitch** [stylelint/stylelint#3259]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Bug where using the `:global` exception in css modules is not excluded from the selector-class-pattern rule. https://github.com/css-modules/css-modules#exceptions.  For example, if you are using some 3rd party css that you want to override, you use `:global` in css modules.  It wasn't possible to ignore selectors with that prefix via the regex.

> Which rule, if any, is this issue related to?

`selector-class-pattern`

> What CSS is needed to reproduce this issue?

e.g.

```css
:global .some-3rd__party {
  color: pink;
}
:global(.another-3rd__party) {
  color: pink;
}
```

> What stylelint configuration is needed to reproduce this issue?

e.g.

```json
{
  "rules": {
    "selector-class-pattern": "^[a-z][a-zA-Z0-9]+$", // camel case selectors
  }
}
```

> Which version of stylelint are you using?

e.g. `9.2.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI with `stylelint '**/*.css'`

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Yes, css modules.

> What did you expect to happen?

Ignore the rule if the global scope is added

> What actually happened (e.g. what warnings or errors you are getting)?

e.g. "The following warnings were flagged:"

```shell
Expected class selector ".some-3rd__party" to match specified pattern         selector-class-pattern
```

**Comment by jeddy3** [comment]
@gpoitch Thanks for the report and for using the template.

I'd say the rule is behaving correctly here as `.some-3rd__party` is a class. You'll want to use the [disable commands](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/configuration.md#turning-rules-off-from-within-your-css) to turn off the rule when you encounter these exceptions.

**Comment by gpoitch** [comment]
@jeddy3 Thanks for having a look.  Disabling per rule is certainly an option.  A quick look at the code had me thinking that this could be an exception as a non-standard rule?: https://github.com/stylelint/stylelint/blob/f6fbad37e1ffa84aed4a996b3829b4275e7320c5/lib/utils/isStandardSyntaxRule.js

**Comment by jeddy3** [comment]
I appreciate you digging into the code :) 

The `isStandardSyntax*` utils are for excluding non-standard constructs introduced in language extensions like Less and SCSS from being categorised by PostCSS as standard constructs e.g. SCSS nested properties and Less guards are categorised as rules by PostCSS. The `isStandardSyntaxRule` excludes these types of constructs so the built-in rules can just focus on standard constructs as supporting all non-standard constructs is not possible.

Where as I believe `:global .some-3rd__party` is a [valid construct](https://www.w3.org/TR/selectors-4/#pseudo-classes) i.e. a _complex selector_ made up of a _pseudo-class selector_ and a _class selector_ separated by a _descendant combinator_. Likewise for `:global(.another-3rd__party)`, which I believe is a _functional pseudo-class selector_ with a _class selector_ as its argument.

So, the `isStandardSyntax` utils aren't the right vehicle for this use case. I'm happy to field queries about other ideas, though.

You could see how much mileage you get from [writing a plugin that builds on top of the rule](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/plugins.md#stylelintrules). However, the recommended approach is to use disable commands in these instances as they explicitly communicate within the code base that there is an exception to the normal.

**Comment by petertwise** [comment]
Just in case anyone else ends up here via google too:

Restrictive: only allow lowercase and dashes (kebab case) in classes and ids:
```
{
  "rules": {
    "selector-class-pattern": "^[a-z][a-z0-9-]+$",
    "selector-id-pattern": "^[a-z][a-z0-9-]+$"
  }
}
```

Additionally allow underscores (snake case) in the mix too:
```
{
  "rules": {
    "selector-class-pattern": "^[a-z][a-z0-9_-]+$",
    "selector-id-pattern": "^[a-z][a-z0-9_-]+$"
  }
}
```

Additionally allow capitals as long as they are not the first character (camel case)
```
{
  "rules": {
    "selector-class-pattern": "^[a-z][a-zA-Z0-9-]+$",
    "selector-id-pattern": "^[a-z][a-zA-Z0-9-]+$"
  }
}
```

Permissive: Allow all types of cases: kebab, snake, camel, mixes
```
{
  "rules": {
    "selector-class-pattern": "^[a-z][a-zA-Z0-9_-]+$",
    "selector-id-pattern": "^[a-z][a-zA-Z0-9_-]+$"
  }
}
```

**Comment by fe-shin** [comment]
and for pascal case with use of BEM (for `--` and `__`):
```
{
  "rules": {
    "selector-class-pattern": "^[A-Z][a-zA-Z0-9_-]+$",
    "selector-id-pattern": "^[A-Z][a-zA-Z0-9_-]+$"
  }
}
```

**Comment by kiwenlau** [comment]
fix "Unexpected unknown pseudo-class selector ":global" (selector-pseudo-class-no-unknown)" by：

```json
{
  "rules": {
    "selector-pseudo-class-no-unknown": [
      true,
      {
        "ignorePseudoClasses": [
          "global"
        ],
      },
    ],
  },
}
```

**Comment by luin** [comment]
Anyone who needs this, here is a custom rule that works the same as selector-class-pattern excepts that it ignores `:global` scope. It also works with nested local overrides that it can catch `.invalid-selector` in  `:global { .name:local(.invalid-selector) {} }`:

```js
const stylelint = require('stylelint');
const isStandardSyntaxRule = require('stylelint/lib/utils/isStandardSyntaxRule');
const isKeyframeSelector = require('stylelint/lib/utils/isKeyframeSelector');
const parseSelector = require('stylelint/lib/utils/parseSelector');

const { report, ruleMessages } = stylelint.utils;
const ruleName = 'slab/selector-class-pattern';
const messages = ruleMessages(ruleName, {
  expected: (unfixed, fixed) => `Expected "${unfixed}" to be "${fixed}"`,
});

module.exports = stylelint.createPlugin(
  ruleName,
  (primary) => (root, result) => {
    const normalizedPattern =
      typeof primary === 'string' ? new RegExp(primary) : primary;

    root.walkRules((ruleNode) => {
      const { selector, selectors } = ruleNode;

      if (!isStandardSyntaxRule(ruleNode)) {
        return;
      }

      if (selectors.some((s) => isKeyframeSelector(s))) {
        return;
      }

      parseSelector(selector, result, ruleNode, (s) =>
        checkSelector(s, ruleNode),
      );
    });

    function checkSelector(selectorNode, ruleNode) {
      selectorNode.walkClasses((classNode) => {
        const { value, sourceIndex: index } = classNode;

        if (normalizedPattern.test(value)) {
          return;
        }

        const scope = closestScope(classNode);
        if (scope === ':global') {
          return;
        }

        if (scope !== ':local' && closestScope(ruleNode) === ':global') {
          return;
        }
        const selector = String(classNode);
        const endIndex = index + selector.length;

        report({
          result,
          ruleName,
          message: messages.expected,
          messageArgs: [selector, primary],
          node: ruleNode,
          index,
          endIndex,
        });
      });
    }
  },
);

const closestScope = (node) => {
  let parent = node.parent;
  while (parent) {
    const value = parent.value || parent.selector;
    if (value === ':global' || value === ':local') return value;
    parent = parent.parent;
  }
};

module.exports.ruleName = ruleName;
module.exports.messages = messages;
```

**Comment by egucciar** [comment]
thank you, @luin for this. I've had chatgpt update it to work with the latest stylelint and also be able to capture classes defined like this:

```
.someModule {
  :global {
     .some-global-class {
      }
  }
}
```
